waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,512 @@
+{.push raises: [].}
+
+import
+  std/[options, sugar, sequtils],
+  stew/byteutils,
+  results,
+  chronicles,
+  chronos,
+  metrics,
+  libp2p/utility,
+  libp2p/protocols/protocol,
+  libp2p/stream/connection,
+  libp2p/crypto/crypto,
+  eth/p2p/discoveryv5/enr
+import
+  ../common/nimchronos,
+  ../common/enr,
+  ../waku_core,
+  ../waku_archive,
+  ../waku_store/[client, common],
+  ../waku_enr,
+  ../node/peer_manager/peer_manager,
+  ./raw_bindings,
+  ./common,
+  ./session
+
+logScope:
+  topics = "waku sync"
+
+type WakuSync* = ref object of LPProtocol
+  storage: NegentropyStorage
+  maxFrameSize: int # Negentropy param to limit the size of payloads
+
+  peerManager: PeerManager
+
+  syncInterval: timer.Duration # Time between each syncronisation attempt
+  syncRange: timer.Duration # Amount of time in the past to sync
+  relayJitter: Duration # Amount of time since the present to ignore when syncing
+  transferCallBack: Option[TransferCallback] # Callback for message transfers.
+
+  pruneCallBack: Option[PruneCallBack] # Callback with the result of the archive query
+  pruneStart: Timestamp # Last pruning start timestamp
+  pruneOffset: timer.Duration # Offset to prune a bit more than necessary. 
+
+  periodicSyncFut: Future[void]
+  periodicPruneFut: Future[void]
+
+proc storageSize*(self: WakuSync): int =
+  self.storage.len
+
+proc messageIngress*(self: WakuSync, pubsubTopic: PubsubTopic, msg: WakuMessage) =
+  if msg.ephemeral:
+    return
+
+  let msgHash: WakuMessageHash = computeMessageHash(pubsubTopic, msg)
+
+  trace "inserting message into waku sync storage ",
+    msg_hash = msgHash.to0xHex(), timestamp = msg.timestamp
+
+  self.storage.insert(msg.timestamp, msgHash).isOkOr:
+    error "failed to insert message ", msg_hash = msgHash.to0xHex(), error = $error
+
+proc messageIngress*(
+    self: WakuSync, pubsubTopic: PubsubTopic, msgHash: WakuMessageHash, msg: WakuMessage
+) =
+  if msg.ephemeral:
+    return
+
+  trace "inserting message into waku sync storage ",
+    msg_hash = msgHash.to0xHex(), timestamp = msg.timestamp
+
+  if self.storage.insert(msg.timestamp, msgHash).isErr():
+    error "failed to insert message ", msg_hash = msgHash.to0xHex()
+
+proc calculateRange(
+    jitter: Duration = 20.seconds, syncRange: Duration = 1.hours
+): (int64, int64) =
+  ## Calculates the start and end time of a sync session
+
+  var now = getNowInNanosecondTime()
+
+  # Because of message jitter inherent to Relay protocol
+  now -= jitter.nanos
+
+  let syncRange = syncRange.nanos
+
+  let syncStart = now - syncRange
+  let syncEnd = now
+
+  return (syncStart, syncEnd)
+
+proc request(
+    self: WakuSync, conn: Connection
+): Future[Result[seq[WakuMessageHash], string]] {.async.} =
+  let (syncStart, syncEnd) = calculateRange(self.relayJitter)
+
+  let initialized =
+    ?clientInitialize(self.storage, conn, self.maxFrameSize, syncStart, syncEnd)
```

Comment:
Since `request()` return a result and is internal only I wanted to avoid double logging.

In cases where `?` is used it is true that the caller is responsible for logging.